src/custom_request_schedule_manager.cc

```diff
@@ -147,4 +150,67 @@ CustomRequestScheduleManager::GetTimestamp(size_t dataset_index) const
   return std::chrono::milliseconds(timestamp_ms);
 }
 
+void
+CustomRequestScheduleManager::PerformWarmup()
+{
+  if (warmup_schedule_.empty()) {
+    throw std::runtime_error("Expected warmup schedule to be non-empty.");
+  }
+
+  const size_t old_max_threads_{max_threads_};
+
+  max_threads_ = std::min(max_threads_, warmup_schedule_.size());
+
+  InitCustomSchedule(warmup_schedule_);
+  WaitForWarmupAndCleanup();
+
+  max_threads_ = old_max_threads_;
+}
+
+void
+CustomRequestScheduleManager::InitCustomSchedule(
+    const Schedule& schedule, size_t dataset_offset)
+{
+  PauseWorkers();
+  const size_t request_count{schedule.size()};
+  ConfigureThreads(request_count, dataset_offset);
+  DistributeScheduleToWorkers(schedule);
+  ResumeWorkers();
+}
+
+std::vector<RateSchedulePtr_t>
+CustomRequestScheduleManager::CreateWorkerSchedules(const Schedule& schedule)
+{
+  std::vector<RateSchedulePtr_t> worker_schedules =
+      CreateEmptyWorkerSchedules();
+  std::vector<size_t> thread_ids{CalculateThreadIds()};
+  size_t thread_id_index = 0;
+  size_t worker_index = 0;
+
+  for (const auto& timestamp : schedule) {
+    const auto timestamp_ns{
+        std::chrono::duration_cast<std::chrono::nanoseconds>(timestamp)};
+    worker_index = thread_ids[thread_id_index];
+    thread_id_index = ++thread_id_index % thread_ids.size();
+    worker_schedules[worker_index]->intervals.push_back(timestamp_ns);
+  }
+  SetScheduleDurations(worker_schedules);
+
+  return worker_schedules;
+}
+
+void
+CustomRequestScheduleManager::StartBenchmark()
+{
+  if (benchmark_schedule_.empty()) {
+    throw std::runtime_error("Expected benchmark schedule to be non-empty.");
+  }
+
+  max_threads_ = std::min(max_threads_, benchmark_schedule_.size());
+
+  const size_t dataset_offset{warmup_schedule_.size()};
+
+  InitCustomSchedule(benchmark_schedule_, dataset_offset);
```

Comment:
This is also part of the new logic for fixed schedule warmup implementation.

Now, after warmup, the workers take a "dataset offset" value to know to skip that many payloads when starting benchmarking.

They skip however many payloads were sent during warmup.